cli/azd/cmd/show.go

```diff
@@ -0,0 +1,188 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/pkg/commands"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment/azdcontext"
+	"github.com/azure/azure-dev/cli/azd/pkg/infra"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/project"
+	"github.com/spf13/cobra"
+)
+
+func showCmd(rootOptions *internal.GlobalCommandOptions) *cobra.Command {
+	action := func(ctx context.Context, cmd *cobra.Command, args []string, azdCtx *azdcontext.AzdContext) error {
+		console := input.GetConsole(ctx)
+
+		formatter := output.GetFormatter(ctx)
+		writer := output.GetWriter(ctx)
+
+		// Right now this command is hidden and we only expect it to be called by tooling,
+		// which passes `--output json`. If for some reason someone ran it directly, just
+		// don't do anything.
+		if formatter.Kind() != output.JsonFormat {
+			return nil
+		}
+
+		if err := ensureProject(azdCtx.ProjectPath()); err != nil {
+			return err
+		}
+
+		env, ctx, err := loadOrInitEnvironment(ctx, &rootOptions.EnvironmentName, azdCtx, console)
+		if err != nil {
+			return fmt.Errorf("loading environment: %w", err)
+		}
+
+		prj, err := project.LoadProjectConfig(azdCtx.ProjectPath(), &env)
+		if err != nil {
+			return fmt.Errorf("loading project: %w", err)
+		}
+
+		res := showResult{
+			Services: make(map[string]showService, len(prj.Services)),
+		}
+
+		for name, svc := range prj.Services {
+			path, err := getFullPathToProjectForService(svc)
+			if err != nil {
+				return err
+			}
+
+			showSvc := showService{
+				Project: showServiceProject{
+					Path: path,
+					Type: showTypeFromLanguage(svc.Language),
+				},
+			}
+
+			res.Services[name] = showSvc
+		}
+
+		// Add information about the target of each service, if we can determine it (if the infrastructure has
+		// not been deployed, for example, we'll just not include target information)
+		resourceManager := infra.NewAzureResourceManager(ctx)
+		if resourceGroupName, err := resourceManager.FindResourceGroupForEnvironment(ctx, &env); err == nil {
+			for name := range prj.Services {
+				if resources, err := project.GetServiceResources(ctx, resourceGroupName, name, &env); err == nil {
+					if len(resources) == 1 {
```

Comment:
Hrm.  This is a good question.  I'm not sure if we would want to fully error out (for example, we ignore other errors on this path, but we do log them at least, so I should add some logging code that says we found multiple and are ignoring the issue).  I think it is helpful to just not include the target and return the rest of the information if we can.  I think there's a lot of information we can communicate even if we can't find the target resource (and since we don't fail when the target doesn't exist, whereas `deploy` would, it feels okay to me that there's a disconnect here).

Another thing I could imagine:  You're using VS to and doing `azd` stuff.  You make a mistake an accidently tag two resources with the same tag (maybe a copy paste error or something during a refactoring) and then do a provision.  While `deploy` would now be broken (at least for that service) I wouldn't want to break the entire editor experience and I worry returning an error here will cause that.

Do you feel okay about just logging for now?  I could also imagine we write some warning to `stderr` (so we don't mess up the JSON output) to the effect of `warning: could not determine the target for service %s, multiple resources were found tagged with azd-service-name: %s.`

But I don't think I want to fail the complete command.

What do you think?

Reply:
Why don't we just return everything we have found? 

Is it a valid assumption that (in the context of a single environment) there is at most one Azure resource that corresponds to a given service project?

Reply:
Ah, I did misread the `GetServiceResource` that it does log and continue if multiple resources are discovered, so keeping it consistent would be the least invasive approach.

I suspect if we truly want a 1:1 relationship, we need to enforce it correctly everywhere as an invariant. Otherwise, we should allow multiple. Would VS break if there were multiple resources found?

Reply:
Yeah - thinking about this more - I think we need to allow a single "azd service" to be deployed to multiple targets.  I suspect at some point we are going to want to support services which are deployed to multiple regions (for example, you have your API service deployed to an app service in a WestUS and another copy in EastUS for HA reasons, and right now `azd depoloy` doesn't handle that case great, which we'll need to go fix.

I think I will change this such that the JSON now looks like this:

```json
      "target": {
        "resourceIds": ["/subscriptions/faa080af-c1d8-40ad-9cce-e1a450ca5b57/resourceGroups/rg-matell-tst-shw/providers/Microsoft.Web/sites/app-t4e7rfh6jsb5u"]
      }
```

If we can-not discover any resources, the array will be empty.  If we discover multiple resources, they'll all be in there.